validphys2/src/validphys/lhapdfset.py

```diff
@@ -0,0 +1,138 @@
+        log.info("PDF: %s ErrorType: %s", name, error_type)
+        self._name = name
+        self._error_type = error_type
+        if self.is_t0:
+            # If at this point we already know this is a T0 set, load only the CV
+            self._lhapdf_set = [lhapdf.mkPDF(name)]
+        else:
+            self._lhapdf_set = lhapdf.mkPDFs(name)
+        self._flavors = None
+
+    @property
+    def is_monte_carlo(self):
+        """Check whether the error type is MC"""
+        return self._error_type == "replicas"
+
+    @property
+    def is_t0(self):
+        """Check whether we are in t0 mode"""
+        return self._error_type == "t0"
+
+    @property
+    def n_members(self):
+        """Return the number of active members in the PDF set"""
+        return len(self.members)
+
+    @property
+    def members(self):
+        """Return the members of the set, this depends on the error type:
+        t0: returns only member 0
+        MC: skip member 0
+        Hessian: return all
+        """
+        if self.is_t0:
+            return self._lhapdf_set[0:1]
+        if self.is_monte_carlo:
+            return self._lhapdf_set[1:]
+        return self._lhapdf_set
+
+    @property
+    def central_member(self):
+        """Returns a reference to member 0 of the PDF list"""
+        return self._lhapdf_set[0]
+
+    def xfxQ(self, x, Q, n, fl):
+        """Return the PDF value for one single point for one single member
+        If the flavour is not included in the PDF (for instance top/antitop) return 0.0
+        """
+        member_pdf = self.members[n]
+        res = member_pdf.xfxQ(x, Q)
+        return res.get(fl, 0.0)
+
+    @property
+    def flavors(self):
+        """Returns the list of accepted flavors by the LHAPDF set"""
+        if self._flavors is None:
+            self._flavors = self.members[0].flavors()
+        return self._flavors
+
+    def grid_values(self, flavors: np.ndarray, xgrid: np.ndarray, qgrid: np.ndarray):
+        """Returns the PDF values for every member for the required
+        flavours, points in x and pointx in q
+        The return shape is
+            (members, flavors, xgrid, qgrid)
+        Return
+        ------
+            ndarray of shape (members, flavors, xgrid, qgrid)
+        Examples
+        --------
+        >>> import numpy as np
+        >>> from validphys.lhapdfset import LHAPDFSet, ER_MC
+        >>> pdf = LHAPDFSet("NNPDF40_nnlo_as_01180", ER_MC)
+        >>> xgrid = np.random.rand(10)
+        >>> qgrid = np.random.rand(3)
+        >>> flavs = np.arange(-4,4)
+        >>> flavs[4] = 21
+        >>> results = pdf.grid_values(flavs, xgrid, qgrid)
+        """
+        # Grid values loop
+        ret = np.array(
+            [
+                [[member.xfxQ(flavors, x, q) for x in xgrid] for q in qgrid]
+                for member in self.members
+            ]
+        )
+
+        # Finally return in the documented shape
+        return np.transpose(ret, (0, 3, 2, 1))
```

Comment:
Aparently you can call LHAPDF with arrays. This seems to be a tad faster:

```python
def gv_2(self, flavors: np.ndarray, xgrid: np.ndarray, qgrid: np.ndarray):
    mxgrid, mqgrid = (g.ravel() for g in np.meshgrid(xgrid, qgrid))
    raw = np.array([mem.xfxQ(flgrid, mxgrid, mqgrid) for mem in self.members])
    return raw.reshape(
        len(self.members), len(qgrid), len(xgrid), len(flgrid)
    ).transpose(0, 3, 2, 1)
```

%timeit gv_2(l_pdf ,flgrid, xgrid, qgrid)
82.6 ms ± 1.65 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

vs
%timeit gv_1(l_pdf ,flgrid, xgrid, qgrid)
108 ms ± 1.28 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

Reply:
oh, you're right. I'm sure I tried this (a wrong variation of, clearly,) and didn't work. I'll update it.

Why did you need to use meshgrid? Would a call with xgrid/qgrid not work out of the box?

Reply:
Because it zips the inputs, which is slightly inconvenient for our use.

Reply:
nvm, I had forgotten what our grid values was doing (which is of course the reason you had to use `meshgrid` there)